# Tugas Praproses

### Kelompok 1

1. Muhammad Fadlul Hafiizh
2. ..

## Pra-Proses Data

In [1]:
%matplotlib inline  
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt  
import seaborn as sns

In [2]:
df_fasilitas_motor=pd.read_excel("dataset/raw_data.xlsx",sheet_name='fasilitas_motor')
df_fasilitas_dump_truck=pd.read_excel("dataset/raw_data.xlsx",sheet_name='fasilitas_dump_truck')
df_fasilitas_arm_truck=pd.read_excel("dataset/raw_data.xlsx",sheet_name='fasilitas_arm_truck')

### Fasilitas
#### Eksplorasi atribut

In [3]:
df_fasilitas_motor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 135 entries, 0 to 134
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   id                   135 non-null    int64 
 1   kode_provinsi        135 non-null    int64 
 2   nama_provinsi        135 non-null    object
 3   kode_kabupaten_kota  135 non-null    int64 
 4   nama_kabupaten_kota  135 non-null    object
 5   jumlah_motor_sampah  135 non-null    int64 
 6   satuan               135 non-null    object
 7   tahun                135 non-null    int64 
dtypes: int64(5), object(3)
memory usage: 8.6+ KB


In [4]:
df_fasilitas_dump_truck.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 135 entries, 0 to 134
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   id                   135 non-null    int64 
 1   kode_provinsi        135 non-null    int64 
 2   nama_provinsi        135 non-null    object
 3   kode_kabupaten_kota  135 non-null    int64 
 4   nama_kabupaten_kota  135 non-null    object
 5   jumlah_dumptruck     135 non-null    int64 
 6   satuan               135 non-null    object
 7   tahun                135 non-null    int64 
dtypes: int64(5), object(3)
memory usage: 8.6+ KB


In [5]:
df_fasilitas_arm_truck.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 135 entries, 0 to 134
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   id                   135 non-null    int64 
 1   kode_provinsi        135 non-null    int64 
 2   nama_provinsi        135 non-null    object
 3   kode_kabupaten_kota  135 non-null    int64 
 4   nama_kabupaten_kota  135 non-null    object
 5   jumlah_armroll       135 non-null    int64 
 6   satuan               135 non-null    object
 7   tahun                135 non-null    int64 
dtypes: int64(5), object(3)
memory usage: 8.6+ KB


#### Drop kolom
Kolom kode provinsi, nama provinsi, nama kabupaten kota, dan satuan didrop karena pada tabel keseluruhan fasilitas, semua berisikan data yang sama yang sudah distandarisasi. Namun, karena tetap memerlukan keterangan tersebut, hanya dua dari tiga tabel fasilitas yang didrop kolomnya.

In [6]:
dropped_attr = ['kode_provinsi', 'nama_provinsi', 'nama_kabupaten_kota', 'satuan']
df_fasilitas_dump_truck = df_fasilitas_dump_truck.drop(dropped_attr,axis=1)
df_fasilitas_arm_truck = df_fasilitas_arm_truck.drop(dropped_attr,axis=1)

df_fasilitas_motor.info()
print("\n")
df_fasilitas_arm_truck.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 135 entries, 0 to 134
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   id                   135 non-null    int64 
 1   kode_provinsi        135 non-null    int64 
 2   nama_provinsi        135 non-null    object
 3   kode_kabupaten_kota  135 non-null    int64 
 4   nama_kabupaten_kota  135 non-null    object
 5   jumlah_motor_sampah  135 non-null    int64 
 6   satuan               135 non-null    object
 7   tahun                135 non-null    int64 
dtypes: int64(5), object(3)
memory usage: 8.6+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 135 entries, 0 to 134
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   id                   135 non-null    int64
 1   kode_kabupaten_kota  135 non-null    int64
 2   jumlah_armroll       135 non-null    int64
 3   tahun

#### Merge tabel
Semua tabel dilakukan merge untuk mendapatkan keseluruhan data fasilitas. Full outer digunakan agar mendapatkan seluruh data, yang ada di tabel. Diketahui bahwa record yang membedakan dari tiap tabel hanyalah jumlah fasilitasnya, sehingga tidak memerlukan praproses lebih lanjut.

In [7]:
df_fasilitas = pd.merge(df_fasilitas_arm_truck, df_fasilitas_dump_truck, how="outer", on=["kode_kabupaten_kota", "tahun"])
df_fasilitas = pd.merge(df_fasilitas, df_fasilitas_motor, how="outer", on=["kode_kabupaten_kota", "tahun"])
df_fasilitas

id_x  kode_kabupaten_kota  jumlah_armroll  tahun  id_y  jumlah_dumptruck  \
0       1                 3201               9   2016     1               136   
1       2                 3202               7   2016     2                23   
2       3                 3203               4   2016     3                25   
3       4                 3204              22   2016     4                59   
4       5                 3205               4   2016     5                14   
..    ...                  ...             ...    ...   ...               ...   
130   131                 3275              78   2021   131               173   
131   132                 3276              20   2021   132                92   
132   133                 3277              24   2021   133                16   
133   134                 3278              12   2021   134                29   
134   135                 3279               6   2021   135                 2   

      id  kode_provinsi nama_provinsi nama_kabupaten_kota  \
0      1             32    JAWA BARAT     KABUPATEN BOGOR   
1      2             32    JAWA BARAT  KABUPATEN SUKABUMI   
2      3             32    JAWA BARAT   KABUPATEN CIANJUR   
3      4             32    JAWA BARAT   KABUPATEN BANDUNG   
4      5             32    JAWA BARAT     KABUPATEN GARUT   
..   ...            ...           ...                 ...   
130  131             32    JAWA BARAT         KOTA BEKASI   
131  132             32    JAWA BARAT          KOTA DEPOK   
132  133             32    JAWA BARAT         KOTA CIMAHI   
133  134             32    JAWA BARAT    KOTA TASIKMALAYA   
134  135             32    JAWA BARAT         KOTA BANJAR   

     jumlah_motor_sampah satuan  
0                     23   UNIT  
1                     12   UNIT  
2                     14   UNIT  
3                     38   UNIT  
4                      0   UNIT  
..                   ...    ...  
130                   41   UNIT  
131                   70   UNIT  
132                  170   UNIT  
133                   30   UNIT  
134                   26   UNIT  

[135 rows x 12 columns]

### Sampah
#### Eksplorasi Atribut

In [8]:
df_produksi_sampah = pd.read_excel("dataset/raw_data.xlsx",sheet_name='jumlah_produksi_sampah')
df_sampah_ditangani = pd.read_excel("dataset/raw_data.xlsx",sheet_name='jumlah_sampah_ditangani')
df_sampah_tertimbun = pd.read_excel("dataset/raw_data.xlsx",sheet_name='sampah_tertimbun')
df_sampah_diangkut = pd.read_excel("dataset/raw_data.xlsx",sheet_name='sampah_terangkut')

df_produksi_sampah.info()
print("\n")
df_sampah_ditangani.info()
print("\n")
df_sampah_tertimbun.info()
print("\n")
df_sampah_diangkut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 189 entries, 0 to 188
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      189 non-null    int64  
 1   kode_provinsi           189 non-null    int64  
 2   nama_provinsi           189 non-null    object 
 3   kode_kabupaten_kota     189 non-null    int64  
 4   nama_kabupaten_kota     189 non-null    object 
 5   jumlah_produksi_sampah  189 non-null    float64
 6   satuan                  189 non-null    object 
 7   tahun                   189 non-null    int64  
dtypes: float64(1), int64(4), object(3)
memory usage: 11.9+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 189 entries, 0 to 188
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   189 non-null    int64  
 1   kode_provinsi        189 non-null    int64

Jika dilihat dari informasi yang terdapat diatas ditambah dengan informasi singkat (penamaan dataset dari opendata.jabarprov.go.id) dan informasi susunan kolom yang mirip kita bisa mendapatkan gambaran konteks/lingkup dari data tersebut. Bahwasanya dari keempat dataset diatas ada kemungkinan data tersebut saling berelasi yang terpisah pisah dan menggambarkan suatu proses perjalanan data, namun masih perlu ditelaah lebih lanjut. Maka dari itu kita coba cocokan terlebih dahulu informasi data didalamnya untuk memastikan bahwasanya data ini memiliki pola yang saling terhubung. Jika kita lihat setiap dataset memiliki sebuah nilai satuan maka dari itu kita coba lihat setiap nilai satuan didalam masing masing dataset

In [9]:
value_count = df_produksi_sampah.satuan.value_counts()
print("Satuan untuk dataset produksi sampah")
print(value_count)

value_count = df_sampah_ditangani.satuan.value_counts()
print("Satuan untuk dataset sampah yang ditangani")
print(value_count)

value_count = df_sampah_tertimbun.satuan.value_counts()
print("Satuan untuk dataset sampah yang tertimbun")
print(value_count)

value_count = df_sampah_diangkut.satuan.value_counts()
print("Satuan untuk dataset sampah yang diangkut")
print(value_count)

Satuan untuk dataset produksi sampah
TON PER HARI    189
Name: satuan, dtype: int64
Satuan untuk dataset sampah yang ditangani
TON PER HARI    189
Name: satuan, dtype: int64
Satuan untuk dataset sampah yang tertimbun
TON PER HARI    135
Name: satuan, dtype: int64
Satuan untuk dataset sampah yang diangkut
TON/HARI    135
Name: satuan, dtype: int64


Jika dilihat dari semua dataset itu, semuanya memiliki jenis satuan yang sama akan tetapi ada kendala inkonsistensi pada
penulisan satuannya dan jumlah baris nya terbagi menjadi dua kelompok ada yang 189 dan 135 baris dan juga diatas pada hasil eksplorasi melihat informasi dari data frame sampah yang diangkut itu ada jumlah value pada kolom jumlah_kolom yang 134 sedangkan yang lainnya berjumlah 135, ini mungkin ada ada jumlah_sampah yang null. OLeh karena itu kita lanjutkan untuk menangani data tersebut


#### Pembersihan & Transformasi

replace satuan yang tidak konsisten menjadi satu nama satuan yang sama

In [10]:
df_sampah_diangkut['satuan'] = df_sampah_diangkut['satuan'].replace(['TON/HARI'], 'TON PER HARI')
df_sampah_diangkut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 135 entries, 0 to 134
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   135 non-null    int64  
 1   kode_provinsi        135 non-null    int64  
 2   nama_provinsi        135 non-null    object 
 3   kode_kabupaten_kota  135 non-null    int64  
 4   nama_kabupaten_kota  135 non-null    object 
 5   jumlah_sampah        134 non-null    float64
 6   satuan               135 non-null    object 
 7   tahun                135 non-null    int64  
dtypes: float64(1), int64(4), object(3)
memory usage: 8.6+ KB


Lalu untuk data yang null pada kolom jumlah sampah dari data frame sampah yang diangkut kita ganti dengan nilai 0.

In [11]:
df_sampah_diangkut.jumlah_sampah.fillna(0, inplace=True)

Selanjutnya kita gabungkan terlebih dahulu dataset diatas untuk mengetahui apa yang menyebabkan jumlah barisnya berbeda.
kita gabungkan dengan metode full outer agar tidak ada data yang berkurang dan dapat melihat data apa yang tidak
lengkap/Nan/Null.

In [21]:
df_pelayanan_sampah = pd.merge(df_produksi_sampah, (df_sampah_ditangani[['kode_kabupaten_kota','tahun','jumlah_sampah']]).rename(columns={'jumlah_sampah':'jumlah_sampah_ditangani'}), how='outer', on=['kode_kabupaten_kota', 'tahun'])
df_pelayanan_sampah = pd.merge(df_pelayanan_sampah, df_sampah_tertimbun[['kode_kabupaten_kota','tahun','timbunan_sampah_kota']], how='outer', on=['kode_kabupaten_kota', 'tahun'])
df_pelayanan_sampah = pd.merge(df_pelayanan_sampah, (df_sampah_diangkut[['kode_kabupaten_kota','tahun','jumlah_sampah']]).rename(columns={'jumlah_sampah':'jumlah_sampah_terangkut'}), how='outer', on=['kode_kabupaten_kota', 'tahun'])
df_pelayanan_sampah = pd.merge(df_pelayanan_sampah, (df_fasilitas[['kode_kabupaten_kota','tahun','jumlah_armroll', 'jumlah_dumptruck', 'jumlah_motor_sampah']]), how='outer', on=['kode_kabupaten_kota', 'tahun'])
df_pelayanan_sampah.info()
df_pelayanan_sampah

<class 'pandas.core.frame.DataFrame'>
Int64Index: 189 entries, 0 to 188
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       189 non-null    int64  
 1   kode_provinsi            189 non-null    int64  
 2   nama_provinsi            189 non-null    object 
 3   kode_kabupaten_kota      189 non-null    int64  
 4   nama_kabupaten_kota      189 non-null    object 
 5   jumlah_produksi_sampah   189 non-null    float64
 6   satuan                   189 non-null    object 
 7   tahun                    189 non-null    int64  
 8   jumlah_sampah_ditangani  189 non-null    float64
 9   timbunan_sampah_kota     135 non-null    float64
 10  jumlah_sampah_terangkut  135 non-null    float64
 11  jumlah_armroll           135 non-null    float64
 12  jumlah_dumptruck         135 non-null    float64
 13  jumlah_motor_sampah      135 non-null    float64
dtypes: float64(7), int64(4), o

id  kode_provinsi nama_provinsi  kode_kabupaten_kota  \
0      1             32    JAWA BARAT                 3201   
1      2             32    JAWA BARAT                 3202   
2      3             32    JAWA BARAT                 3203   
3      4             32    JAWA BARAT                 3204   
4      5             32    JAWA BARAT                 3205   
..   ...            ...           ...                  ...   
184  185             32    JAWA BARAT                 3275   
185  186             32    JAWA BARAT                 3276   
186  187             32    JAWA BARAT                 3277   
187  188             32    JAWA BARAT                 3278   
188  189             32    JAWA BARAT                 3279   

    nama_kabupaten_kota  jumlah_produksi_sampah        satuan  tahun  \
0       KABUPATEN BOGOR                 1511.15  TON PER HARI   2015   
1    KABUPATEN SUKABUMI                  419.01  TON PER HARI   2015   
2     KABUPATEN CIANJUR                  981.41  TON PER HARI   2015   
3     KABUPATEN BANDUNG                 1895.94  TON PER HARI   2015   
4       KABUPATEN GARUT                  464.74  TON PER HARI   2015   
..                  ...                     ...           ...    ...   
184         KOTA BEKASI                 1500.77  TON PER HARI   2021   
185          KOTA DEPOK                 1418.87  TON PER HARI   2021   
186         KOTA CIMAHI                  193.36  TON PER HARI   2021   
187    KOTA TASIKMALAYA                  295.87  TON PER HARI   2021   
188         KOTA BANJAR                   60.29  TON PER HARI   2021   

     jumlah_sampah_ditangani  timbunan_sampah_kota  jumlah_sampah_terangkut  \
0                  861919.30                   NaN                      NaN   
1                  205690.20                   NaN                      NaN   
2                  355554.30                   NaN                      NaN   
3                 1156523.00                   NaN                      NaN   
4                  237389.80                   NaN                      NaN   
..                       ...                   ...                      ...   
184                  1061.00               1500.77                  1061.00   
185                  1194.20               1418.87                  1194.20   
186                   173.80                193.36                   173.80   
187                   194.29                295.87                   194.29   
188                    44.00                 60.29                    44.00   

     jumlah_armroll  jumlah_dumptruck  jumlah_motor_sampah  
0               NaN               NaN                  NaN  
1               NaN               NaN                  NaN  
2               NaN               NaN                  NaN  
3               NaN               NaN                  NaN  
4               NaN               NaN                  NaN  
..              ...               ...                  ...  
184            78.0             173.0                 41.0  
185            20.0              92.0                 70.0  
186            24.0              16.0                170.0  
187            12.0              29.0                 30.0  
188             6.0               2.0                 26.0  

[189 rows x 14 columns]

Jika dilihat 